### PR TITLE
moved logic for normalization of fileName from Issue constructor to I…

### PR DESCRIPTION
**Pull request by nengelbrecht**
…ssuBuilder#build.

changed type of parameter 'fileName' from String to TreeString within Issue (constructor and fileName-setter)

### CODE REVIEW

File: src/main/java/edu/hm/hafner/analysis/Issue.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 @SuppressWarnings({"PMD.TooManyFields", "PMD.GodClass", "NoFunctionalReturnType"})
 public class Issue implements Serializable {
     private static final long serialVersionUID = 1L; // release 1.0.0
-    private static final String UNDEFINED = "-";
+    protected static final String UNDEFINED = "-";
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/Issue.java]: package private should be sufficient.

File: src/main/java/edu/hm/hafner/analysis/IssueBuilder.java [hunk 2]
```diff
@@ -28,6 +30,9 @@
  */
 @SuppressWarnings({"InstanceVariableMayNotBeInitialized", "JavaDocMethod", "PMD.TooManyFields"})
 public class IssueBuilder {
+
+    private static TreeStringBuilder treeStringBuilder = new TreeStringBuilder();
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/IssueBuilder.java]: While caching at class level is most effective I think that might be too risky at the moment (memory leak, synchronization). During parsing of an analysis report only a single instance of `IssueBuilder` is used so it should be sufficient to use a `TreeStringBuilder` per instance. So remove `static` for the moment, this is something we can tweak later on.

File: src/main/java/edu/hm/hafner/analysis/IssueBuilder.java [hunk 3]
```diff
@@ -224,13 +229,34 @@ public IssueBuilder copy(final Issue copy) {
...
     public Issue build() {
-        Issue issue = new Issue(fileName, lineStart, lineEnd, columnStart, columnEnd, lineRanges, category, type,
+        Issue issue = new Issue(treeStringOfFileName(fileName), lineStart, lineEnd, columnStart, columnEnd, lineRanges, category, type,
                 packageName, moduleName, severity, message, description, origin, reference, fingerprint,
                 additionalProperties, id);
         id = UUID.randomUUID(); // make sure that multiple invocations will create different IDs
         return issue;
     }
 
+    protected static TreeString treeStringOfFileName(@Nullable final String fileName) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/IssueBuilder.java]: - package private
- add `@VisibleForTesting`

File: src/test/java/edu/hm/hafner/analysis/IssueBuilderTest.java [hunk 4]
```diff
@@ -19,9 +19,9 @@
  * @author Marcel Binder
  */
 class IssueBuilderTest {
-    private static final Issue DEFAULT_ISSUE = new Issue(null, 0, 0, 0, 0, new LineRangeList(),
+    private static final Issue DEFAULT_ISSUE = new Issue(UNDEFINED_TS, 0, 0, 0, 0, new LineRangeList(),
             null, null, null, null, null, null, null, null, null, null, null);
-    private static final Issue FILLED_ISSUE = new Issue(FILE_NAME, LINE_START, LINE_END, COLUMN_START, COLUMN_END,
+    private static final Issue FILLED_ISSUE = new Issue(IssueBuilder.treeStringOfFileName(FILE_NAME), LINE_START, LINE_END, COLUMN_START, COLUMN_END,
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/IssueBuilderTest.java]: In order to test the caching, you can add a test like this:
```
    @Test
    void shouldCacheFileName() {
        IssueBuilder builder = new IssueBuilder();

        Issue issue = builder.setFileName("fileName").build();
        Issue anotherIssue = builder.setFileName("fileName").build();

        assertThat(issue.getFileNameTreeString()).isSameAs(anotherIssue.getFileNameTreeString());
    }
```

This test should fail for the other properties but already works for the `fileName`.
Not that you need to create a package private @VisibleForTesting method `getFileNameTreeString` in Issue.

**Comment by uhafner** [comment]
The CI build currently fails, since this change is an API breaking change. Please upgrade the version in the pom to 5.0.0-SNAPSHOT than the build should succeed.

**Comment by uhafner** [comment]
I updated `equals` and `hashCode` of `TreeString` in https://github.com/jenkinsci/analysis-model/commit/7d2bcace1d9767330553af2e22e263a207bbf5c1.

**Comment by nengelbrecht** [comment]
I did not change the parameter-type of Issue#setPackageName yet because PackageNameResolver#run uses the setter. This could be a problem due to moving the transformation-logic to the IssueBuilder class. (When TreeStringBuilder would be static, the methods for transformation could be static as well and be used by Issue#setPackageName in order to avoid redundancy)

**Comment by uhafner** [comment]
I merged it into a separate branch, see https://github.com/jenkinsci/analysis-model/tree/treestring.
So I have the opportunity to evolve the Issue API even more.

**Pull request closed without merging**